metrics-api/src/main/java/software/amazon/swage/metrics/MetricContext.java

```diff
@@ -73,6 +77,24 @@ default void count(Metric label) {
         count(label, 1L);
     }
 
+    /**
+     * Create a child context from this instance which is differentiated by the supplied attributes.
+     * A child context can be used to add additional attributes as information about the
+     * environment is produced or discovered.
+     * For example, an application may create a child context for processing a request at the time that the request is received.
+     *
+     * @param attributes The attributes that identify the child context in relation to this instance.
+     * @return a new child context of this instance.
+     */
+    MetricContext newChild(TypedMap attributes);
+
+    /**
+     * Returns this instance's parent. The parent contains the attributes which
+     * describe the environment that this {MetricContext} was created from.
+     * @return The {MetricContext} that this instance was created from or null if no parent exists.
+     */
+    MetricContext parent();
```

Comment:
Is there a need for the parent to be explicit in the interface?

I can see implementations where a MetricContext may carry around pointers to parent(s) and reconstruct the full context, and other implementations where the context is built from the parent at start and no pointers are kept.  What is gained by making access to the parent explicit in the interface?

Reply:
I would argue that flattening should be the responsibility of the recorder and recorder context. MetricContexts by definition are hierarchical. I don't see a reason that a MetricContext would be making decisions about what information the recorder needs. 

I see two benefits for this being part of the interface. The first is that it makes the hierarchical nature of MetricContext part of the definition, which makes reasoning about the context within an application easier. Second, it makes managing contexts easier. Libraries that manage the context within an application such as ThreadLocalMetrics, do not need to maintain their own hierarchy that mirrors that of the MetricContext or rely on a specific subclass of MetricContexts.